### PR TITLE
MQE: add split operator

### PR DESCRIPTION
**Pull request by charleskorn**
#### What this PR does

This PR adds a split operator and node, equivalent in functionality to the splitting logic in the split and cache frontend middleware.

My idea is that there will be two operators: one for splitting the range into smaller ranges and combining the results together, and another that is responsible for caching the result of a sub-range. The split node and cache node would be injected at the root of the query plan by a new optimisation pass.

The cache node and optimisation pass described above will come in a later PR. This operator is not yet used anywhere, I'm opening a PR with just these changes in the interests of making it easier to review. 

Given this change is not user-visible, I've chosen not to add a changelog entry at this time.

Note that the operator does not emit the same metrics as the existing middleware - I will address this in a later PR.

#### Which issue(s) this PR fixes or relates to

(none)

#### Checklist

- [x] Tests updated.
- [n/a] Documentation added.
- [n/a] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`. If changelog entry is not needed, please add the `changelog-not-needed` label to the PR.
- [n/a] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

---

> [!NOTE]
> **Medium Risk**
> Adds a new query plan node/operator that changes how range queries can be executed (split into sub-ranges) and bumps the query plan version, which could affect correctness/perf once enabled despite being unused by default.
> 
> **Overview**
> Introduces a new `TimeRangeSplit` query-plan node and `TimeRangeSplitOperator` that can **split a range query into interval-aligned sub-ranges**, execute each sub-range operator, and then merge series metadata, samples (floats + histograms), annotations, and per-step evaluation stats.
> 
> Wires the new node type into planning/serialization (`NODE_TYPE_TIME_RANGE_SPLIT`), bumps plan version to `QueryPlanV14`, and registers a materializer in the streaming PromQL engine. Adds `OperatorEvaluationStats.AddSubRange()` to combine stats from sub-range evaluations, plus test-only operator enhancements and comprehensive unit tests for splitting/merging behavior and edge cases (eg. conflicting `DropName`).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 8dec4ded8ee0440e259a8793080fae6c032544d7. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: pkg/streamingpromql/optimize/plan/splitandcache/split_operator.go [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+			labelBytesBuf = series.Labels.Bytes(labelBytesBuf)
+
+			if outputSeriesIdx, seenAlready := seriesIndices[string(labelBytesBuf)]; seenAlready {
+				if series.DropName != allSeries[outputSeriesIdx].DropName {
+					return nil, fmt.Errorf("series with labels %s has conflicting drop name values in different ranges", series.Labels.String())
+				}
+
+				s.outputSeries[outputSeriesIdx].sourceSeriesIndices[rangeIdx] = rangeSeriesIdx
+
+				// We're not doing to keep this labels instance (we already have it from a previous range), so
```
Comment (56quarters) [inline pkg/streamingpromql/optimize/plan/splitandcache/split_operator.go]:
```suggestion
				// We're not going to keep this labels instance (we already have it from a previous range), so
```

File: pkg/streamingpromql/optimize/plan/splitandcache/split_operator.go [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+		return nil, err
+	}
+
+	s.ranges[0].buffer = operators.NewInstantVectorOperatorBuffer(s.ranges[0].operator, nil, len(allSeries)-1, s.MemoryConsumptionTracker)
+
+	seriesIndices := make(map[string]int, len(allSeries))
+	labelBytesBuf := make([]byte, 0, 1024)
+	for seriesIdx, series := range allSeries {
+		labelBytesBuf = series.Labels.Bytes(labelBytesBuf)
+		seriesIndices[string(labelBytesBuf)] = seriesIdx
```
Comment (56quarters) [inline pkg/streamingpromql/optimize/plan/splitandcache/split_operator.go]: Can you add a comment here about why we do `string(bytes)` inline like this?

File: pkg/streamingpromql/optimize/plan/splitandcache/split_operator.go [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+
+	thisOutputSeries := s.outputSeries[0]
+	s.outputSeries = s.outputSeries[1:]
+
+	defer types.IntSlicePool.Put(&thisOutputSeries.sourceSeriesIndices, s.MemoryConsumptionTracker)
+
+	var floats []promql.FPoint
+	var histograms []promql.HPoint
+
+	for rangeIdx, sourceSeriesIdx := range thisOutputSeries.sourceSeriesIndices {
```
Comment (56quarters) [inline pkg/streamingpromql/optimize/plan/splitandcache/split_operator.go:176]: Can you document on this type that the source series are indexed by range?

File: pkg/streamingpromql/optimize/plan/splitandcache/split_operator.go [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+	"github.com/prometheus/prometheus/promql/parser/posrange"
+
+	"github.com/grafana/mimir/pkg/querier/stats"
+	"github.com/grafana/mimir/pkg/streamingpromql/operators"
+	"github.com/grafana/mimir/pkg/streamingpromql/types"
+	"github.com/grafana/mimir/pkg/util/limiter"
+)
+
+// TimeRangeSplitOperator evaluates a range query by concatenating the results of multiple shorter-range queries.
+type TimeRangeSplitOperator struct {
```
Comment (56quarters) [inline pkg/streamingpromql/optimize/plan/splitandcache/split_operator.go:20]: Can you add a type assertion for the appropriate operator interface for this type?

**Review by 56quarters — DISMISSED** [review]
LGTM with a few minor comments.

File: pkg/streamingpromql/optimize/plan/splitandcache/split_operator.go (hunk 3 above)
Comment (charleskorn) [reply pkg/streamingpromql/optimize/plan/splitandcache/split_operator.go:176]: The definition of `splitOutputSeries` already has this comment, is that what you mean? 

```go
type splitOutputSeries struct {
	sourceSeriesIndices []int // One entry per range. -1 indicates that the series is not present in the range.
}
```

File: pkg/streamingpromql/optimize/plan/splitandcache/split_operator.go (hunk 1 above)
Comment (charleskorn) [reply pkg/streamingpromql/optimize/plan/splitandcache/split_operator.go]: Good catch, fixed in 67ff926.

File: pkg/streamingpromql/optimize/plan/splitandcache/split_operator.go (hunk 2 above)
Comment (charleskorn) [reply pkg/streamingpromql/optimize/plan/splitandcache/split_operator.go]: Added in 67ff926

File: pkg/streamingpromql/optimize/plan/splitandcache/split_operator.go (hunk 4 above)
Comment (charleskorn) [reply pkg/streamingpromql/optimize/plan/splitandcache/split_operator.go:20]: Added in 67ff926.

**Review by tcp13equals2 — APPROVED** [review]
